tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -0,0 +1,266 @@
+        optimization_guard_(false),
+        allocation_order_guard_(false) {
+    NVF_CHECK(E % H == 0);
+    NVF_CHECK(H % D == 0);
+    NVF_CHECK(E % D == 0);
+  }
+
+  void SetUp() {
+    MultiDeviceTest::SetUp();
+    if (!deviceMajorMinorCheck(8)) {
+      GTEST_SKIP() << "Distributed transformer tests require Ampere or newer";
+    }
+  }
+
+  hir::HostIrExecutorParams executor_params_{
+      .use_fusion_executor_cache = true,
+      .skip_auto_scheduling = false,
+      .cache_fusion_executor = false};
+  const int D;
+  const int B = 4;
+  const int E = 128;
+  const int H = 4;
+  const int S = 32;
+
+  // Temporary until https://github.com/NVIDIA/Fuser/issues/2563
+  at::Tensor shardTensor(at::Tensor tensor, int64_t axis, DeviceMesh& mesh) {
+    auto i = mesh.idxOf(communicator_->deviceId());
+    auto extent = tensor.size(axis);
+    auto nslices = mesh.size();
+    NVF_CHECK(
+        extent % nslices == 0,
+        "Sharded axis must be evenly divisble by mesh");
+    auto stride = extent / nslices;
+    // TODO: returning slice 0 temporarily when device is not in the mesh.
+    i = (i < 0) ? 0 : i;
+    return tensor.slice(axis, i * stride, (i + 1) * stride)
+                 .contiguous()
+                 .unsqueeze(0);
+  }
+
+ private:
+  // Note: `MoveSplitCat` and `AllocationDomain` preseg passes use ID model.
+  // `SdpaFwdOp` currently does not work with ID model since it requires all
+  // sibling outputs to have the same root domain.
+  //  This will be modified in a future PR.
+  preseg_passes::OptimizationPassGuard<preseg_passes::MoveSplitCatPass>
+      optimization_guard_;
+  preseg_passes::OptimizationPassGuard<preseg_passes::AllocationDomainPass>
+      allocation_order_guard_;
+};
+
+namespace {
+TensorView* replicated_dropout(
+    TensorView* x,
+    const double kProb,
+    Fusion* fusion,
+    DeviceMesh mesh) {
+  // Need to modify two things before we can use the existing dropout function
+  // in composite.cpp (1) Sharding propagation breaks at rand_like because it
+  // creates a fresh TV. (2) The philox seed and offset must be set to ensure
+  // the masks are identical across processes.
+  TensorView* x_float = castOp(DataType::Float, x);
+  const double kScale = 1.0 / (1.0 - kProb);
+  Val* philox_seed = fusion->zeroVal();
+  Val* philox_offset = fusion->zeroVal();
+  TensorView* rand_vals = rand_like(x_float, philox_seed, philox_offset);
+  TensorView* mask = lt(rand_vals, IrBuilder::create<Val>(1.0 - kProb));
+  TensorView* apply_mask = mul(x_float, mask);
+  TensorView* dropout = mul(apply_mask, IrBuilder::create<Val>(kScale));
+  rand_vals->setDeviceMesh(mesh);
+  return dropout;
+}
+
+void validate(
+    std::vector<at::Tensor> expected_out,
+    std::vector<at::Tensor> out) {
+  EXPECT_EQ(expected_out.size(), out.size());
+  for (auto i : c10::irange(out.size())) {
+    // Note: Scale the tolerance up since the error accumulates across ops
+    double tolerance = 0.5 * (i + 1);
```

Comment:
Can you clarify the formula used here?

Reply:
It's quite basic, it's just scales the tolerance up. I picked 0.5 so that the existing tests pass. It's not a precise formula since it's difficult to calculate what half precision error.

Reply:
Got it!